noir-projects/aztec-nr/aztec/src/keys/getters.nr

```diff
@@ -1,79 +1,98 @@
-use dep::protocol_types::{
-  address::{
-    AztecAddress,
-    PartialAddress
-  }, 
-  constants::{
-    GENERATOR_INDEX__PUBLIC_KEYS_HASH,
-    GENERATOR_INDEX__CONTRACT_ADDRESS_V1,
-    CANONICAL_KEY_REGISTRY_ADDRESS
-  },
-  grumpkin_point::GrumpkinPoint,
+use dep::protocol_types::{address::AztecAddress, constants::CANONICAL_KEY_REGISTRY_ADDRESS, grumpkin_point::GrumpkinPoint};
+use crate::{
+    context::PrivateContext, hash::poseidon2_hash, oracle::keys::get_public_keys_and_partial_address,
+    state_vars::{
+    map::derive_storage_slot_in_map,
+    shared_mutable::shared_mutable_private_getter::SharedMutablePrivateGetter
+}
 };
 
-use crate::context::PrivateContext;
-use crate::hash::{
-  pedersen_hash, 
-  poseidon2_hash,
-};
-use crate::oracle::keys::get_public_keys_and_partial_address;
-use crate::state_vars::{
-  map::derive_storage_slot_in_map,
-  shared_mutable::shared_mutable_private_getter::SharedMutablePrivateGetter,
-};
+// Note: In fetch_key_from_registry we expect that shared mutable slow is index * 2 + 1 for x coordinate
+// and index * 2 + 2 for y coordinate, if this changes the function will need to be refactored.
+global NULLIFIER_INDEX = 0;
+global INCOMING_INDEX = 1;
+global OUTGOING_INDEX = 2;
+global TAGGING_INDEX = 3;
+
+global DELAY = 5;
 
-struct PublicKeyTypeEnum {
-  NULLIFIER: u8,
+pub fn get_npk_m(context: &mut PrivateContext, address: AztecAddress) -> GrumpkinPoint {
+    get_master_key(context, address, NULLIFIER_INDEX)
 }
 
-global PublicKeyType = PublicKeyTypeEnum {
-  NULLIFIER: 0,
-};
+pub fn get_ivpk_m(context: &mut PrivateContext, address: AztecAddress) -> GrumpkinPoint {
+    get_master_key(context, address, INCOMING_INDEX)
+}
 
-pub fn get_fresh_nullifier_public_key_hash(
-  context: &mut PrivateContext,
-  address: AztecAddress, 
-) -> Field {
-  // This is the storage slot of the nullifier_public_key inside the key registry contract
-  // TODO: (#6133) We should have this be directly imported from the other contract if possible, or at least this should not be this brittle
-  let storage_slot_of_nullifier_public_key = 1;
+// Commented out as it's currently not enabled in key registry
+// pub fn get_ovpk_m(context: &mut PrivateContext, address: AztecAddress) -> GrumpkinPoint {
+//     get_master_key(context, address, OUTGOING_INDEX)
+// }
+// 
+// pub fn get_tpk_m(context: &mut PrivateContext, address: AztecAddress) -> GrumpkinPoint {
+//     get_master_key(context, address, TAGGING_INDEX)
+// }
 
-  let derived_slot = derive_storage_slot_in_map(storage_slot_of_nullifier_public_key, address);
+fn get_master_key(
+    context: &mut PrivateContext,
+    address: AztecAddress,
+    key_index: Field
+) -> GrumpkinPoint {
+    // 1. Get key hash from registry
+    let key = fetch_key_from_registry(context, key_index, address);
+    if key.is_zero() {
+        // i. Keys were not registered yet --> fetch key from PXE
+        fetch_and_constrain_keys(address)[key_index]
```

Comment:
Sorry, ghosted you on this one and today it randomly appeared in my mind. I had no idea to what is going to happen here so I bother Maxim and doing that was a good decision:

<img width="1027" alt="image" src="https://github.com/AztecProtocol/aztec-packages/assets/13470840/a31ec5f8-ca70-4a2f-a95d-b2781b2e2ef4">

<img width="1024" alt="image" src="https://github.com/AztecProtocol/aztec-packages/assets/13470840/8da8e4c5-c7a1-4c07-8baf-26fd0308aa4a">

So basically noir can afford to not execute the oracle call because in that situation it inserts 0 to the constraint and then it can fill in the return values from the oracle with any bullshit.